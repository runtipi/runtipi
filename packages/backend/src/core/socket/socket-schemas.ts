import { z } from 'zod';

export const socketEventSchema = z.union([
  z.object({
    type: z.literal('app'),
    event: z.union([
      z.literal('status_change'),
      z.literal('install_success'),
      z.literal('install_error'),
      z.literal('uninstall_success'),
      z.literal('uninstall_error'),
      z.literal('reset_success'),
      z.literal('reset_error'),
      z.literal('update_success'),
      z.literal('update_error'),
      z.literal('start_success'),
      z.literal('start_error'),
      z.literal('stop_success'),
      z.literal('stop_error'),
      z.literal('restart_success'),
      z.literal('restart_error'),
      z.literal('generate_env_success'),
      z.literal('generate_env_error'),
      z.literal('backup_success'),
      z.literal('backup_error'),
      z.literal('restore_success'),
      z.literal('restore_error'),
    ]),
    data: z.object({
      appUrn: z.string(),
      appStatus: z
        .enum([
          'running',
          'stopped',
          'starting',
          'stopping',
          'updating',
          'missing',
          'installing',
          'uninstalling',
          'resetting',
          'restarting',
          'backing_up',
          'restoring',
        ])
        .optional(),
      error: z.string().optional(),
    }),
  }),
  z.object({
    type: z.literal('app-logs-init'),
    event: z.literal('initLogs'),
    data: z.object({
      appUrn: z.string(),
      maxLines: z.number().optional(),
    }),
  }),
  z.object({
    type: z.literal('app-logs'),
    event: z.union([z.literal('newLogs'), z.literal('stopLogs')]),
    data: z.object({
      appUrn: z.string(),
      lines: z.array(z.string()).optional(),
    }),
  }),
  z.object({
    type: z.literal('runtipi-logs-init'),
    event: z.literal('initLogs'),
    data: z.object({
      maxLines: z.number().optional(),
    }),
  }),
  z.object({
    type: z.literal('runtipi-logs'),
    event: z.union([z.literal('newLogs'), z.literal('stopLogs')]),
    data: z.object({
      lines: z.array(z.string()).optional(),
    }),
  }),
]);

export type SocketEvent = z.infer<typeof socketEventSchema>;
