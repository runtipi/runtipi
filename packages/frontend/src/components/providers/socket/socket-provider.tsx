import { useSocket } from '@/lib/hooks/use-socket';
import { useQueryClient } from '@tanstack/react-query';
import type { PropsWithChildren } from 'react';
import toast from 'react-hot-toast';
import { useTranslation } from 'react-i18next';

export const SocketProvider = ({ children }: PropsWithChildren) => {
  const { t } = useTranslation();
  const queryClient = useQueryClient();

  useSocket({
    onEvent: (event, data) => {
      if (data.error) {
        console.error(data.error);
      }

      queryClient.invalidateQueries();

      switch (event) {
        case 'status_change':
          break;
        case 'install_success':
          toast.success(t('APP_INSTALL_SUCCESS', { id: data.appId }));
          break;
        case 'install_error':
          toast.error(t('APP_ERROR_APP_FAILED_TO_INSTALL', { id: data.appId }));
          break;
        case 'start_success':
          toast.success(t('APP_START_SUCCESS', { id: data.appId }));
          break;
        case 'start_error':
          toast.error(t('APP_ERROR_APP_FAILED_TO_START', { id: data.appId }));
          break;
        case 'stop_success':
          toast.success(t('APP_STOP_SUCCESS', { id: data.appId }));
          break;
        case 'stop_error':
          toast.error(t('APP_ERROR_APP_FAILED_TO_STOP', { id: data.appId }));
          break;
        case 'uninstall_success':
          toast.success(t('APP_UNINSTALL_SUCCESS', { id: data.appId }));
          break;
        case 'uninstall_error':
          toast.error(t('APP_ERROR_APP_FAILED_TO_UNINSTALL', { id: data.appId }));
          break;
        case 'update_success':
          toast.success(t('APP_UPDATE_SUCCESS', { id: data.appId }));
          break;
        case 'update_error':
          toast.error(t('APP_ERROR_APP_FAILED_TO_UPDATE', { id: data.appId }));
          break;
        case 'reset_success':
          toast.success(t('APP_RESET_SUCCESS', { id: data.appId }));
          break;
        case 'reset_error':
          toast.error(t('APP_ERROR_APP_FAILED_TO_RESET', { id: data.appId }));
          break;
        case 'restart_success':
          toast.success(t('APP_RESTART_SUCCESS', { id: data.appId }));
          break;
        case 'restart_error':
          toast.error(t('APP_ERROR_APP_FAILED_TO_RESTART', { id: data.appId }));
          break;
        case 'backup_success':
          toast.success(t('APP_BACKUP_SUCCESS', { id: data.appId }));
          break;
        case 'backup_error':
          toast.error(t('APP_BACKUP_ERROR', { id: data.appId }));
          break;
        case 'restore_success':
          toast.success(t('APP_RESTORE_SUCCESS', { id: data.appId }));
          break;
        case 'restore_error':
          toast.error(t('APP_RESTORE_ERROR', { id: data.appId }));
          break;
        default:
          break;
      }
    },
    selector: { type: 'app' },
  });

  return children;
};
