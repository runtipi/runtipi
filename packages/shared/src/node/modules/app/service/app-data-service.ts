import { notEmpty } from '../../../helpers/typescript-helpers';
import { DataAccessApp } from '../data-access/data-access-app';

export interface IAppDataService {
  getAppInfoFromInstalledOrAppStore: AppDataService['getAppInfoFromInstalledOrAppStore'];
  getInstalledInfo: AppDataService['getInstalledInfo'];
  getInfoFromAppStore: AppDataService['getInfoFromAppStore'];
  getUpdateInfo: AppDataService['getUpdateInfo'];
  getAllAvailableApps: AppDataService['getAllAvailableApps'];
  getAppBackups: AppDataService['getAppBackups'];
  deleteAppBackup: AppDataService['deleteAppBackup'];
}

export class AppDataService implements IAppDataService {
  private dataAccessApp: DataAccessApp;

  constructor(params: { dataDir: string; appDataDir: string; appsRepoId: string }) {
    const { dataDir, appDataDir, appsRepoId } = params;
    this.dataAccessApp = new DataAccessApp({ dataDir, appDataDir, appsRepoId });
  }

  /**
   *  This function reads the config.json and metadata/description.md files for the app with the provided id,
   *  parses the config file and returns an object with app information.
   *  It checks if the app is installed or not and looks for the config.json file in the appropriate directory.
   *  If the config.json file is invalid, it returns null.
   *  If an error occurs during the process, it logs the error message and throws an error.
   *
   *  @param {string} id - The app id.
   */
  public async getAppInfoFromInstalledOrAppStore(id: string) {
    const info = await this.dataAccessApp.getInstalledAppInfo(id);
    if (!info) {
      return this.dataAccessApp.getAppInfoFromAppStore(id);
    }
    return info;
  }

  public getInstalledInfo(id: string) {
    return this.dataAccessApp.getInstalledAppInfo(id);
  }

  public getInfoFromAppStore(id: string) {
    return this.dataAccessApp.getAppInfoFromAppStore(id);
  }

  public async getUpdateInfo(id: string) {
    return this.dataAccessApp.getAppUpdateInfo(id);
  }

  public async getAllAvailableApps() {
    const appIds = await this.dataAccessApp.getAvailableAppIds();

    const apps = await Promise.all(
      appIds.map(async (app) => {
        return this.dataAccessApp.getAppInfoFromAppStore(app);
      }),
    );

    return apps.filter(notEmpty).map(({ id, categories, name, short_desc, deprecated, supported_architectures, created_at }) => ({
      id,
      categories,
      name,
      short_desc,
      deprecated,
      supported_architectures,
      created_at,
    }));
  }

  public async getAppBackups(params: { appId: string; pageSize: number; page: number }) {
    const { appId, page, pageSize } = params;
    const backups = await this.dataAccessApp.listBackupsByAppId(appId);
    backups.sort((a, b) => b.date.getTime() - a.date.getTime());

    const start = (page - 1) * pageSize;
    const end = start + pageSize;
    const data = backups.slice(start, end);

    return {
      data,
      total: backups.length,
      currentPage: Math.floor(start / pageSize) + 1,
      lastPage: Math.ceil(backups.length / pageSize),
    };
  }

  public async deleteAppBackup(appId: string, filename: string) {
    return this.dataAccessApp.deleteBackup(appId, filename);
  }
}
