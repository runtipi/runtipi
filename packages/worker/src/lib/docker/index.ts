import { compose, handleViewAppLogsEvent, handleViewRuntipiLogsEvent } from './docker-helpers';

export { compose, handleViewAppLogsEvent, handleViewRuntipiLogsEvent };
