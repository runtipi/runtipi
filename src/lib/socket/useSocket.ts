import { type SocketEvent, socketEventSchema } from '@runtipi/shared';
import { useEffect, useRef, useState } from 'react';
import io, { type Socket } from 'socket.io-client';

// Data selector is used to select a specific property/value from the data object if it exists
type DataSelector<T> = {
  property: keyof Extract<SocketEvent, { type: T }>['data'];
  value: unknown;
};

type Selector<T, U> = {
  type: T;
  event?: U;
  data?: DataSelector<T>;
};

type Props<T, U> = {
  onEvent?: (event: Extract<Extract<SocketEvent, { type: T }>['event'], U>, data: Extract<SocketEvent, { type: T }>['data']) => void;
  onError?: (error: string) => void;
  onCleanup?: () => void;
  emitOnConnect?: SocketEvent;
  emitOnDisconnect?: SocketEvent;
  initialData?: Extract<SocketEvent, { type: T }>['data'] | undefined;
  selector: Selector<T, U>;
};

export const useSocket = <T extends SocketEvent['type'], U extends SocketEvent['event']>(props: Props<T, U>) => {
  const { onEvent, onError, onCleanup, selector, initialData, emitOnConnect, emitOnDisconnect } = props;
  const [lastData, setLastData] = useState(initialData as unknown);
  const socketRef = useRef<Socket>();

  // biome-ignore lint/correctness/useExhaustiveDependencies: This effect should never re-run
  useEffect(() => {
    const { hostname, port, protocol } = window.location;

    if (!socketRef.current) {
      socketRef.current = io(`${protocol}//${hostname}:${port}`, { path: '/worker/socket.io' });
    }

    if (socketRef.current?.disconnected) {
      socketRef.current.connect();
    }

    socketRef.current.on('connect', () => {
      if (emitOnConnect) {
        socketRef.current?.emit(emitOnConnect.type, emitOnConnect);
      }
    });

    const handleEvent = (type: SocketEvent['type'], rawData: unknown) => {
      const parsedEvent = socketEventSchema.safeParse(rawData);

      if (!parsedEvent.success) {
        return;
      }

      const { event, data } = parsedEvent.data;

      if (selector) {
        if (selector.type !== type) {
          return;
        }

        if (selector.event && selector.event !== event) {
          return;
        }

        const property = selector.data?.property as keyof SocketEvent['data'];
        if (selector.data && data && selector.data.value !== data[property]) {
          return;
        }
      }

      setLastData(data);
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore - This is fine
      if (onEvent) onEvent(event, data);
    };

    socketRef.current?.on(selector.type as string, (data) => {
      handleEvent(selector.type, data);
    });

    socketRef.current?.on('error', (error: string) => {
      onError?.(String(error));
    });

    return () => {
      if (emitOnDisconnect) {
        socketRef.current?.emit(emitOnDisconnect.type, emitOnDisconnect);
      }

      socketRef.current?.off(selector.type as string);
      socketRef.current?.off('error');
      socketRef.current = undefined;
      onCleanup?.();
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps -- This effect should never re-run
  }, []);

  return { lastData, socket: socketRef.current } as { lastData: Extract<SocketEvent, { type: T }>['data'] | undefined; socket: Socket | undefined };
};
