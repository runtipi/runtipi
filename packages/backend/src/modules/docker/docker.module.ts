import { SocketModule } from '@/core/socket/socket.module';
import { Module } from '@nestjs/common';
import { AppStoreModule } from '../app-stores/app-store.module';
import { AppsModule } from '../apps/apps.module';
import { DockerService } from './docker.service';

@Module({
  imports: [AppsModule, AppStoreModule, SocketModule],
  providers: [DockerService],
  exports: [DockerService],
})
export class DockerModule {}
