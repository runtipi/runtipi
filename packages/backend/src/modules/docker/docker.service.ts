import path from 'node:path';
import { extractAppId } from '@/common/helpers/app-helpers';
import { execAsync } from '@/common/helpers/exec-helpers';
import { ConfigurationService } from '@/core/config/configuration.service';
import { FilesystemService } from '@/core/filesystem/filesystem.service';
import { LoggerService } from '@/core/logger/logger.service';
import { SocketManager } from '@/core/socket/socket.service';
import { Injectable } from '@nestjs/common';
import { AppFilesManager } from '../apps/app-files-manager';
import { DockerCommandFactory } from './docker-command.factory';

@Injectable()
export class DockerService {
  dockerCommandFactory: DockerCommandFactory;

  constructor(
    private readonly logger: LoggerService,
    private readonly config: ConfigurationService,
    private readonly appFilesManager: AppFilesManager,
    private readonly socketManager: SocketManager,
    private readonly filesystem: FilesystemService,
  ) {
    this.dockerCommandFactory = new DockerCommandFactory(this);

    const io = this.socketManager.init();

    io.on('connection', async (socket) => {
      socket.onAny((event, body) => {
        const command = this.dockerCommandFactory.createCommand(event);

        if (command) {
          command.execute(socket, body, this.socketManager.emit.bind(this.socketManager));
        }
      });
    });
  }

  /**
   * Get the base compose args for an app
   * @param {string} appId - App name
   */
  public getBaseComposeArgsApp = async (appId: string) => {
    let isCustomConfig = false;
    const { directories } = this.config.getConfig();

    const { storeId } = extractAppId(appId);

    const appEnv = await this.appFilesManager.getAppEnv(appId);
    const args: string[] = [`--env-file ${appEnv.path}`];

    // User custom env file
    const userEnvFile = await this.appFilesManager.getUserEnv(appId);
    if (userEnvFile.content) {
      isCustomConfig = true;
      args.push(`--env-file ${userEnvFile.path}`);
    }

    args.push(`--project-name ${appId}`);

    const composeFile = await this.appFilesManager.getDockerComposeYaml(appId);
    args.push(`-f ${composeFile.path}`);

    const commonComposeFile = path.join(directories.dataDir, 'repos', storeId, 'apps', 'docker-compose.common.yml');
    args.push(`-f ${commonComposeFile}`);

    // User defined overrides
    const userComposeFile = await this.appFilesManager.getUserComposeFile(appId);
    if (userComposeFile.content) {
      isCustomConfig = true;
      args.push(`--file ${userComposeFile.path}`);
    }

    return { args, isCustomConfig };
  };

  public getBaseComposeArgsRuntipi = async () => {
    const { dataDir } = this.config.get('directories');
    const args: string[] = [`--env-file ${path.join(dataDir, '.env')}`];

    args.push('--project-name runtipi');

    const composeFile = path.join(dataDir, 'docker-compose.yml');
    args.push(`-f ${composeFile}`);

    // User defined overrides
    const userComposeFile = path.join(dataDir, 'user-config', 'tipi-compose.yml');
    if (await this.filesystem.pathExists(userComposeFile)) {
      args.push(`--file ${userComposeFile}`);
    }

    return args;
  };

  /**
   * Helpers to execute docker compose commands
   * @param {string} appId - App name
   * @param {string} command - Command to execute
   */
  public composeApp = async (appId: string, command: string) => {
    const { args, isCustomConfig } = await this.getBaseComposeArgsApp(appId);
    args.push(command);

    this.logger.info(`Running docker compose with args ${args.join(' ')}`);
    const { stdout, stderr } = await execAsync(`docker-compose ${args.join(' ')}`);

    if (stderr?.includes('Command failed:')) {
      if (isCustomConfig) {
        throw new Error(`Error with your custom app: ${stderr}`);
      }

      throw new Error(stderr);
    }

    return { stdout, stderr };
  };
}
