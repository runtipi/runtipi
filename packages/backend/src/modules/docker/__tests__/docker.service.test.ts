import { SocketManager } from '@/core/socket/socket.service';
import { Test } from '@nestjs/testing';
import { fromPartial } from '@total-typescript/shoehorn';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import { mock } from 'vitest-mock-extended';
import type { ServiceInput } from '../builders/schemas';
import { DockerService } from '../docker.service';

describe('DockerService', () => {
  let dockerService: DockerService;
  const mockSocketManager = mock<SocketManager>(fromPartial({ init: () => ({ on: vi.fn() }) }));

  beforeEach(async () => {
    const moduleRef = await Test.createTestingModule({
      providers: [DockerService, { provide: SocketManager, useValue: mockSocketManager }],
    })
      .useMocker(mock)
      .compile();

    dockerService = moduleRef.get(DockerService);
  });

  it('should be able to parse a compose.json file', async () => {
    const composeJson: { services: ServiceInput[] } = {
      services: [
        {
          // @ts-expect-error
          something: 'crazy',
          name: 'ctfd',
          image: 'ctfd/ctfd:3.7.5',
          isMain: true,
          internalPort: 8000,
          environment: {
            UPLOAD_FOLDER: '/var/uploads',
            DATABASE_URL: 'mysql+pymysql://tipi:${CTFD_MYSQL_DB_PASSWORD}@ctfd-db/ctfd',
          },
          dependsOn: ['ctfd-db'],
          volumes: [
            {
              hostPath: '${APP_DATA_DIR}/data/uploads',
              containerPath: '/var/log/CTFd',
            },
            {
              hostPath: '${APP_DATA_DIR}/data/uploads',
              containerPath: '/var/uploads',
            },
          ],
        },
        {
          name: 'ctfd-db',
          image: 'mariadb:10.4.12',
          environment: {
            MYSQL_ROOT_PASSWORD: '${CTFD_MYSQL_ROOT_PASSWORD}',
            MYSQL_USER: 'tipi',
            MYSQL_PASSWORD: '${CTFD_MYSQL_DB_PASSWORD}',
            MYSQL_DATABASE: 'ctfd',
          },
          volumes: [
            {
              hostPath: '${APP_DATA_DIR}/data/db',
              containerPath: '/var/lib/mysql',
            },
          ],
          command: ['mysqld', '--character-set-server=utf8mb4', '--collation-server=utf8mb4_unicode_ci', '--wait_timeout=28800', '--log-warnings=0'],
        },
        {
          name: 'ctfd-redis',
          image: 'redis:4',
          volumes: [
            {
              hostPath: '${APP_DATA_DIR}/data/redis',
              containerPath: '/data',
            },
          ],
        },
      ],
    };

    const yaml = dockerService.getDockerCompose(composeJson.services, { appId: 'test-app', openPort: true });

    expect(yaml).toMatchSnapshot();
  });
});
