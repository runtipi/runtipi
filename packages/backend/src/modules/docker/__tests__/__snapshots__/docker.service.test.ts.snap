// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DockerService > should be able to parse a compose.json file 1`] = `
"services:
  ctfd:
    image: ctfd/ctfd:3.7.5
    container_name: ctfd
    restart: unless-stopped
    networks:
      - tipi_main_network
    environment:
      UPLOAD_FOLDER: /var/uploads
      DATABASE_URL: mysql+pymysql://tipi:\${CTFD_MYSQL_DB_PASSWORD}@ctfd-db/ctfd
    ports:
      - \${APP_PORT}:8000
    volumes:
      - \${APP_DATA_DIR}/data/uploads:/var/log/CTFd
      - \${APP_DATA_DIR}/data/uploads:/var/uploads
    depends_on:
      - ctfd-db
    labels:
      generated: true
      traefik.enable: false
      traefik.http.middlewares.ctfd-web-redirect.redirectscheme.scheme: https
      traefik.http.services.ctfd.loadbalancer.server.port: "8000"
  ctfd-db:
    image: mariadb:10.4.12
    command:
      - mysqld
      - --character-set-server=utf8mb4
      - --collation-server=utf8mb4_unicode_ci
      - --wait_timeout=28800
      - --log-warnings=0
    container_name: ctfd-db
    restart: unless-stopped
    networks:
      - tipi_main_network
    environment:
      MYSQL_ROOT_PASSWORD: \${CTFD_MYSQL_ROOT_PASSWORD}
      MYSQL_USER: tipi
      MYSQL_PASSWORD: \${CTFD_MYSQL_DB_PASSWORD}
      MYSQL_DATABASE: ctfd
    volumes:
      - \${APP_DATA_DIR}/data/db:/var/lib/mysql
  ctfd-redis:
    image: redis:4
    container_name: ctfd-redis
    restart: unless-stopped
    networks:
      - tipi_main_network
    volumes:
      - \${APP_DATA_DIR}/data/redis:/data
networks:
  tipi_main_network:
    name: runtipi_tipi_main_network
    external: true
"
`;
